core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala

```diff
@@ -97,13 +97,40 @@ class MockWorker(master: RpcEndpointRef, conf: SparkConf = new SparkConf) extend
   }
 }
 
-class MockExecutorLaunchFailWorker(master: RpcEndpointRef, conf: SparkConf = new SparkConf)
-  extends MockWorker(master, conf) {
+// This class is designed to handle the lifecycle of only one application.
+class MockExecutorLaunchFailWorker(master: Master, conf: SparkConf = new SparkConf)
+  extends MockWorker(master.self, conf) with Eventually {
+
+  val appRegistered = new CountDownLatch(1)
+  val launchExecutorReceived = new CountDownLatch(1)
+  val appIdsToLaunchExecutor = new mutable.HashSet[String]
   var failedCnt = 0
+
   override def receive: PartialFunction[Any, Unit] = {
+    case LaunchDriver(driverId, desc, resources_) =>
+      drivers += driverId
+      driverResources(driverId) = resources_.map(r => (r._1, r._2.addresses.toSet))
+      master.self.send(RegisterApplication(appDesc, newDriver(driverId)))
+
+      // Below code doesn't make driver stuck, as newDriver opens another rpc endpoint for
+      // handling driver related messages. It guarantees registering application is done
+      // before handling LaunchExecutor message.
+      eventually(timeout(10.seconds)) {
```

Comment:
There're two different event dispatchers in MockExecutorLaunchFailWorker; 'worker' and 'driver' (once it receives `LaunchDriver` message) which we shouldn't assume messages will be handled sequentially across event dispatchers, LaunchExecutor and RegisteredApplication in this case.

That's why I just inject verification code here; this would block handling LaunchExecutor until we verify application registering is done successfully.

Reply:
> this would block handling LaunchExecutor until we verify application registering is done successfully.

How does this block handling `LaunchExecutor`? I believe `Worker` is not a thread safe `RpcEndpoint`. Do I miss something?

Reply:
Yeah you're right my bad. I assumed that's ThreadSafeRpcEndpoint and it's not. Just fixed via applying count down latch between them.